Add Custom Interface Support in Spring Data JPA

In original Spring Data JPA, it supports custom interface to extend the standard interface.

Inherit custom PostRepository. 

```java
public interface PostRepository extends JpaRepository<Post, String>
, PostRepositoryCustom{

}
```
Define a custom interface.

```java
public interface PostRepositoryCustom {
    List<Post> findByKeyword(String q, int page, int size);
}
```
Implements the custom interface.

```java
public class PostRepositoryImpl implements PostRepositoryCustom {
    @PersistenceContext
    EntityManager entityManager;

    @Override
    public List<Post> findByKeyword(String q, int page, int size) {
        CriteriaBuilder cb = this.entityManager.getCriteriaBuilder();
        CriteriaQuery<Post> query = cb.createQuery(Post.class);
        Root<Post> root = query.from(Post.class);
        if (!StringUtils.isEmpty(q)) {
            query.where(
                    cb.or(
                            cb.like(root.get(Post_.title), "%" + q + "%"),
                            cb.like(root.get(Post_.content), "%" + q + "%")
                    )
            );

        }
        return this.entityManager.createQuery(query)
                .setFirstResult(page * size)
                .setMaxResults(size)
                .getResultList();
    }
}
```

In your codes, just call it from `PostRepository` directly.

```java
  List<Post> posts = this.postRepository.findByKeyword(keyword, page, size);
```

In the Spring Data pattern, the `Custom` postfix is a must(can be changed), and the `Impl` class does not required an extra bean declaration, no `@Component`, `@Singleton` on it. 

I Changed my `PostRepositoryImpl` , add a `@Component`, `@Singleton` on it

```bash
11:15:21,455 ERROR [io.qua.dev.DevModeMain] Failed to start quarkus: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:126)
        at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:180)
        at io.quarkus.dev.DevModeMain.start(DevModeMain.java:94)
        at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl1.<clinit>(ApplicationImpl1.zig:349)
        ... 9 more
Caused by: java.lang.NoSuchMethodError: 'void com.example.PostRepositoryImpl_Bean.<init>(io.quarkus.arc.InjectableInterceptor)'
        at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Default_ComponentsProvider.zig:126)
        at io.quarkus.arc.ArcContainerImpl.<init>(ArcContainerImpl.java:86)
        at io.quarkus.arc.Arc.initialize(Arc.java:18)
        at io.quarkus.arc.runtime.ArcRecorder.getContainer(ArcRecorder.java:33)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources15.deploy_0(ArcProcessor$generateResources15.zig:72)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources15.deploy(ArcProcessor$generateResources15.zig:36)
        at io.quarkus.runner.ApplicationImpl1.<clinit>(ApplicationImpl1.zig:311)
        ... 9 more
```

--- a/docs/src/main/asciidoc/spring-data-jpa-guide.adoc
+++ b/docs/src/main/asciidoc/spring-data-jpa-guide.adoc
@@ -452,11 +452,8 @@ The implementation of that fragment looks like this:
 ----
 import java.util.List;
 
-import org.springframework.stereotype.Component;
-
 import io.quarkus.hibernate.orm.panache.runtime.JpaOperations;
 
-@Component
 public class PersonFragmentImpl implements PersonFragment {
 
     @Override

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java
@@ -21,8 +21,8 @@ import org.springframework.data.repository.PagingAndSortingRepository;
 import org.springframework.data.repository.Repository;
 import org.springframework.data.repository.query.QueryByExampleExecutor;
 
+import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
-import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
@@ -50,14 +50,14 @@ public class SpringDataJPAProcessor {
     @BuildStep
     void build(CombinedIndexBuildItem index,
             BuildProducer<GeneratedBeanBuildItem> generatedBeans,
-            BuildProducer<UnremovableBeanBuildItem> unremovableBeans) {
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
 
         IndexView indexIndex = index.getIndex();
         List<ClassInfo> interfacesExtendingCrudRepository = getAllInterfacesExtending(DotNames.SUPPORTED_REPOSITORIES,
                 indexIndex);
 
         removeNoRepositoryBeanClasses(interfacesExtendingCrudRepository);
-        implementCrudRepositories(generatedBeans, unremovableBeans, interfacesExtendingCrudRepository, indexIndex);
+        implementCrudRepositories(generatedBeans, additionalBeans, interfacesExtendingCrudRepository, indexIndex);
     }
 
     private void removeNoRepositoryBeanClasses(List<ClassInfo> interfacesExtendingCrudRepository) {
@@ -95,7 +95,7 @@ public class SpringDataJPAProcessor {
 
     // generate a concrete class that will be used by Arc to resolve injection points
     private void implementCrudRepositories(BuildProducer<GeneratedBeanBuildItem> generatedBeans,
-            BuildProducer<UnremovableBeanBuildItem> unremovableBeans,
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans,
             List<ClassInfo> crudRepositoriesToImplement, IndexView index) {
 
         ClassOutput classOutput = new ClassOutput() {
@@ -116,7 +116,8 @@ public class SpringDataJPAProcessor {
         CompositeIndex compositeIndex = CompositeIndex.create(index, indexer.complete());
 
         SpringDataRepositoryCreator repositoryCreator = new SpringDataRepositoryCreator(classOutput, compositeIndex, (n) -> {
-            unremovableBeans.produce(new UnremovableBeanBuildItem(new UnremovableBeanBuildItem.BeanClassNameExclusion(n)));
+            // the implementation of fragments don't need to be beans themselves
+            additionalBeans.produce(AdditionalBeanBuildItem.unremovableOf(n));
         });
 
         for (ClassInfo crudRepositoryToImplement : crudRepositoriesToImplement) {

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/FragmentMethodsAdder.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/FragmentMethodsAdder.java
@@ -34,12 +34,27 @@ public class FragmentMethodsAdder {
             List<DotName> customInterfaceNamesToImplement, Map<String, FieldDescriptor> customImplNameToHandle) {
         for (DotName customInterfaceToImplement : customInterfaceNamesToImplement) {
             Collection<ClassInfo> knownImplementors = index.getAllKnownImplementors(customInterfaceToImplement);
-            if (knownImplementors.size() != 1) {
+            ClassInfo customInterfaceImpl = null;
+            if (knownImplementors.size() > 1) {
+                for (ClassInfo knownImplementor : knownImplementors) {
+                    if (knownImplementor.name().toString().endsWith("Impl")) { // the default suffix that Spring Data JPA looks for is 'Impl'
+                        if (customInterfaceImpl != null) { // make sure we don't have multiple implementations suffixed with 'Impl'
+                            throw new IllegalArgumentException(
+                                    "Interface " + customInterfaceToImplement
+                                            + " must contain a single implementation whose name ends with 'Impl'. Multiple implementations were found: "
+                                            + customInterfaceImpl + "," + knownImplementor);
+                        }
+                        customInterfaceImpl = knownImplementor;
+                    }
+                }
+
+            } else if (knownImplementors.size() == 1) {
+                customInterfaceImpl = knownImplementors.iterator().next();
+            } else {
                 throw new IllegalArgumentException(
-                        "Interface " + customInterfaceToImplement + " must contain a single implementation which is a bean");
+                        "No implementation of interface " + customInterfaceToImplement + " was found");
             }
 
-            ClassInfo customInterfaceImpl = knownImplementors.iterator().next();
             String customImplementationClassName = customInterfaceImpl.name().toString();
             fragmentImplClassResolvedCallback.accept(customImplementationClassName);
 

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PersonFragment3Impl.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PersonFragment3Impl.java
@@ -1,8 +1,5 @@
 package io.quarkus.it.spring.data.jpa;
 
-import javax.inject.Singleton;
-
-@Singleton
 public class PersonFragment3Impl implements PersonFragment3 {
 
     @Override

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PersonFragmentImpl.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PersonFragmentImpl.java
@@ -2,17 +2,17 @@ package io.quarkus.it.spring.data.jpa;
 
 import java.util.List;
 
-import org.springframework.stereotype.Component;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
 
-import io.quarkus.hibernate.orm.panache.runtime.JpaOperations;
-
-@Component
 public class PersonFragmentImpl implements PersonFragment, PersonFragment2 {
+
+    @PersistenceContext
+    EntityManager entityManager;
 
     @Override
     public List<Person> findAll() {
-        System.out.println("Custom findAll");
-        return (List<Person>) JpaOperations.findAll(Person.class).list();
+        return entityManager.createQuery("SELECT p FROM Person p").getResultList();
     }
 
     @Override

--- a/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/PersonResourceTest.java
+++ b/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/PersonResourceTest.java
@@ -23,6 +23,15 @@ public class PersonResourceTest {
             Arrays.asList("Bob", "Florence", "DeMar", null));
 
     @Test
+    void testFindAll() {
+        when().get("/person/all").then()
+                .statusCode(200)
+                .body(containsString("DeMar"))
+                .body(containsString("Florence"))
+                .body(containsString("Bob"));
+    }
+
+    @Test
     void testCount() {
         String count = when().get("/person/count").then()
                 .statusCode(200)

